tests/mod_api/test_controller.py

```diff
@@ -898,3 +898,167 @@ def tearDown(self):
+        self.assertEqual(response.status_code, 200)
+
+    @patch('app.mod_api.models.BitStore.copy_to_new_version')
+    @patch('app.mod_api.models.MetaDataDB.create_or_update_version')
+    def test_throw_400_if_version_missing(self, create_or_update_version, copy_to_new_version):
+        copy_to_new_version.return_value = True
+        create_or_update_version.return_value = True
+        response = self.client.post(self.url,
+                                    data=json.dumps({
+                                        'version_fail': 'tag_one'
+                                    }),
+                                    content_type='application/json',
+                                    headers=dict(Authorization=self.auth))
+        self.assertEqual(response.status_code, 400)
+        data = json.loads(response.data)
+        self.assertEqual('ATTRIBUTE_MISSING', data['error_code'])
+
+    @patch('app.mod_api.models.BitStore.copy_to_new_version')
+    @patch('app.mod_api.models.MetaDataDB.create_or_update_version')
+    def test_throw_500_if_failed_to_tag(self, create_or_update_version, copy_to_new_version):
+        copy_to_new_version.return_value = False
+        create_or_update_version.return_value = True
+        response = self.client.post(self.url,
+                                    data=json.dumps({
+                                        'version': 'tag_one'
+                                    }),
+                                    content_type='application/json',
+                                    headers=dict(Authorization=self.auth))
+        self.assertEqual(response.status_code, 500)
+
+    @patch('app.mod_api.models.BitStore.copy_to_new_version')
+    @patch('app.mod_api.models.MetaDataDB.create_or_update_version')
+    def test_throw_403_if_not_owner_or_member_of_publisher(self, create_or_update_version,
+                                                           copy_to_new_version):
+        copy_to_new_version.return_value = False
+        create_or_update_version.return_value = True
+        response = self.client.post(self.jwt_url,
+                                    data=json.dumps({
+                                        'username': self.user_not_allowed_name,
+                                        'secret': 'super_secret'
+                                    }),
+                                    content_type='application/json')
+        data = json.loads(response.data)
+        jwt_not_allowed = data['token']
+        auth_not_allowed = "bearer %s" % jwt_not_allowed
+
+        response = self.client.post(self.url,
+                                    data=json.dumps({
+                                        'version': 'tag_one'
+                                    }),
+                                    content_type='application/json',
+                                    headers=dict(Authorization=auth_not_allowed))
+        self.assertEqual(response.status_code, 403)
```

Comment:
You mocked out the actual data-writing methods. This is a road to introduce security bugs this test should protect us against. For instance if the controller logic is wrong, it could return 403 but still modify the database. In the permissions test it would be desired to check not only the status returned, but that the actual database objects are not changed. This means you should not mock-out these methods at this high level and let em do the work. To check that s3 is intact, the boto client could be MagicMocked, and asserted that ithe mock method was not called to modify the data. To check that a database is intact, you can just query for its current MetadataDB objects and assert that those are left intact, as they was before the test started.